### DISCUSSION TITLE
Verify invoices on smartphone after browsing via a potentially-compromised computer

### DISCUSSION THREAD

**Discussion by ozomer** [btcpayserver/btcpayserver#2421] [Ideas / Feature Requests]
Browsing websites from a PC is more convenient than browsing from a smartphone, but smartphones are considered more secure while PCs can be compromised. A simple example could be a malicious chrome extension that detects you have been browsing to a BTCPay-Server payment page, and redirects the browser to a malicious BTCPay-Server owned by the hacker (with similar payment details).
Even if the user opens the invoice page on his phone and checks the address, it is very typical to have the BTCPay-Server run on a different domain, so for example if the store origin was https://tesla.com , a payment server at https://tesla-payments.com wouldn't look suspicious.

What I want to create is a standard protocol, so that in addition to the "Pay with Bitcoin" and "Pay with Bitcoin (Lightning)" options, the user will have an option of "Verify & Pay with BTCPay Verifier App".
This will show the user a dedicated QR code that contains the BTCPay-Server address (call it `btcpayServerDomain`), the origin address of the store (call it `origin`) and the invoice-id (call it `invoiceId`), and could be scanned by a dedicated smartphone app.

The app will then call "https://`origin`/btcpay-verifier?btcpayServerDomain=...&invoiceId=...", and will expect a json that contains the `btcpayServerAddress` - A simple store website that fully controls the btcpay-server could simply implement this with a static json file like: `{ "domains": ["tesla-payments.com"], "description": "..." }`. A more complicated scenario could analyze the query parameters and return specific details (in case of multiple btcpay-servers and multiple sub-stores).

If the `btcpayPayServerDomain` appears in the "https://`origin`/btcpay-verifier" json, it means the origin trusts this btcpay-server (this is somewhat similar to CORS mechanism).

The smartphone app will then ask the user whether he trusts the origin page. i.e: "**tesla.com** has asked for a payment - please check the spelling of this domain", followed by the payment `description`. It could also be nice to specifically mark with colors any non-english characters that could be used for phishing (and even digits and dashes).

If the user approves, the app will show him the standard invoice page to complete the payment, on the same smartphone or on a third device.

Would love to hear feedback. I can contribute to the development of this feature as well (depends on the amount of free time).

P.S.
This feature is not intended for any doomsday scenario where we could not trust "https".

**Comment by bolatovumar** [comment]
Sounds interesting.

**Comment by woutersamaey** [comment]
Can’t we make a payment page that checks the referer? If the referer matches a fixed url that was mentioned during the invoice creation, you could only reach the payment page going through that door.

Not sure how the referrer header behaves during different kinds of redirects though...

Another solution is not to redirect to a payment page, but serve it on the page itself via a popup...
